src/rmkit/fb/fb.cpy

```diff
@@ -314,12 +318,13 @@ namespace framebuffer:
 
           if src[i] != alpha:
             if image.channels >= 3:
-              ptr[i] = to_rgb565((char *) src, i*image.channels)
+              self._set_pixel(&ptr[i], i, j, to_rgb565((char *) src, i*image.channels))
```

Comment:
If `_set_pixel` was overloaded to take a float or an 8-bit grayscale value, this could produce a higher-quality 16-color dither. But that probably means making `self.dither` a class pointer instead of a function pointer.

Reply:
what's the use case for higher quality dither? i think i have trouble telling 16 grays apart reliably (without patterns) - would this make it easier?

Reply:
It's mostly to reduce banding. You can't see it in the demo app (b/c of using rgb565 internally), but you should be able to see a little less banding in the dithered version. Like you said, it's much more noticeable in a pattern or gradient, but you can see it a little bit in a photograph as well. [This](https://en.wikipedia.org/wiki/List_of_monochrome_and_RGB_color_formats) wikipedia article has an example of the same parrot in 4-bit (16-color) grayscale [without dithering](https://en.wikipedia.org/wiki/File:Grayscale_4bit_palette_sample_image.png) and [with dithering](https://en.wikipedia.org/wiki/File:Grayscale_4bit_palette_sample_image_-_gimp_dithered.png) -- it's _very_ subtle, especially with such a small image, but if you look closely there's some banding in the middle of the bird's chest that's smoothed out in the dithered version.

But yeah, the difference is way more dramatic with 2-color than it will ever be in 16-color, even if we were using floats. And honestly I don't have a need for 16-color dithering at this point, it was more an interesting diversion. I can see why koreader and plato have gone for it though, since it probably makes a big difference for book covers for instance.